utilities/scale-json/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+/// [Value](serde_json::Value).
+///
+/// See the documentation of [ScaleDecodedToJson](crate::ScaleDecodedToJson) for more information.
+#[derive(Debug)]
+struct RawJsonVisitor<R>(PhantomData<R>);
+
+impl<R> Copy for RawJsonVisitor<R> {}
+impl<R> Clone for RawJsonVisitor<R> {
+	fn clone(&self) -> Self {
+		*self
+	}
+}
+impl<R> Default for RawJsonVisitor<R> {
+	fn default() -> Self {
+		Self(Default::default())
+	}
+}
+
+fn trimmed_hex<T: AsRef<[u8]>>(bytes: T) -> String {
+	let s = hex::encode(bytes.as_ref());
+	let trimmed = s.trim_start_matches('0');
+	format!("0x{}", if trimmed.is_empty() { "0" } else { trimmed })
+}
+
+fn untrimmed_hex<T: AsRef<[u8]>>(bytes: T) -> String {
+	format!("0x{}", hex::encode(bytes.as_ref()))
+}
+
+macro_rules! decode_to_number {
+	( $(
+		$name:ident, $t:ty
+	),+ $(,)? ) => {
+		$(
+			fn $name<'scale, 'resolver>(
+				self,
+				value: $t,
+				_type_id: scale_decode::visitor::TypeIdFor<Self>,
+			) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				Ok(serde_json::Value::Number(serde_json::Number::from(value)))
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_hex {
+	(
+		$(
+			$name:ident, $t:ty $( | $convert:ident )?
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+				self,
+				value: $t,
+				_type_id: scale_decode::visitor::TypeIdFor<Self>,
+			) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				Ok(serde_json::Value::String(trimmed_hex(value$( .$convert() )?)))
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_array {
+	(
+		$(
+			$name:ident, $scale_type:ident
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+			self,
+			values: &mut scale_decode::visitor::types::$scale_type<'scale, 'resolver, Self::TypeResolver>,
+			_type_id: scale_decode::visitor::TypeIdFor<Self>,
+		) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				Ok(serde_json::Value::Array(
+					values.map(|res| {
+						res.and_then(|item| {
+							item.decode_with_visitor(self)
+						})
+					}).collect::<Result<Vec<_>,_>>()?
+				))
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_array_or_hex {
+	(
+		$(
+			$name:ident, $scale_type:ident
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+			self,
+			values: &mut scale_decode::visitor::types::$scale_type<'scale, 'resolver, Self::TypeResolver>,
+			_type_id: scale_decode::visitor::TypeIdFor<Self>,
+		) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				// Heuristic to determine if the array is a byte array:
+				// - Array must be non-empty
+				// - AND Each value decodes to an unsigned json Number (ie. not an enum)
+				// - AND each value was encoded as a single byte
+				let (is_bytes, all_values) = values.try_fold(
+					(values.remaining() > 0, Vec::with_capacity(values.remaining())),
+					|(all_bytes, mut values), res| {
+						res.and_then(|item| {
+							let decoded_item = item.decode_with_visitor(self)?;
+							let is_byte = item.bytes().len() == 1 && decoded_item.is_u64();
```

Comment:
not sure I follow what this is about. We check that the item has 1 byte, but it's decoded into 8 bytes? Why is a single byte value not decoded into a u8 here? i.e. why are we not doing `&& decoded_item.is_u8()`

Reply:
`is_u8` doesn't exist unfortunately 😓 .
`decoded_item` here is a `serde_json::Value`. So this is checking if the original scale value was encoded in a single byte *and* that it was decoded to a json number. (actually I need to double-check the single-byte assumption, ironically enough I don't think the assumption holds for compact encoding...).

Think of `is_u64` as meaning `is_unsigned_integer`.

Reply:
Update - pretty sure the compact encoding thing is fine. Arrays/Vecs of bytes are not encoded using compact encoding, which is exactly what we expect. I imagine it's possible for someone to explicitly encode an array of compact-encoded bytes, and in that case, they would be decoded to a json array, not a hex string. I think this is fine: Compact encoding is only useful for numbers wider than a u8.